Sources/SwiftDocC/Semantics/DirectiveInfrastructure/AutomaticDirectiveConvertible.swift

```diff
@@ -0,0 +1,306 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import Markdown
+
+protocol AutomaticDirectiveConvertible: DirectiveConvertible, Semantic {
+    init(originalMarkup: BlockDirective)
+    
+    func validate(
+        source: URL?,
+        for bundle: DocumentationBundle,
+        in context: DocumentationContext,
+        problems: inout [Problem]
+    ) -> Bool
+    
+    static var keyPaths: [String : AnyKeyPath] { get }
+}
+
+extension AutomaticDirectiveConvertible {
+    public static var directiveName: String {
+        String(describing: self)
+    }
+    
+    func validate(
+        source: URL?,
+        for bundle: DocumentationBundle,
+        in context: DocumentationContext,
+        problems: inout [Problem]
+    ) -> Bool {
+        return true
+    }
+}
+
+extension AutomaticDirectiveConvertible {
+    public init?(
+        from directive: BlockDirective,
+        source: URL? = nil,
+        for bundle: DocumentationBundle,
+        in context: DocumentationContext
+    ) {
+        var problems = [Problem]()
+        
+        self.init(
+            from: directive,
+            source: source,
+            for: bundle,
+            in: context,
+            problems: &problems
+        )
+    }
+    
+    public init?(
+        from directive: BlockDirective,
+        source: URL?,
+        for bundle: DocumentationBundle,
+        in context: DocumentationContext,
+        problems: inout [Problem]
+    ) {
+        precondition(directive.name == Self.directiveName)
+        self.init(originalMarkup: directive)
+        
+        let reflectedDirective = DirectiveIndex.shared.reflection(of: type(of: self))
```

Comment:
I haven't run performance tests yet, but I don't expect any because of the `DirectiveIndex` that caches each mirror creation so we only need to make one per directive.

Reply:
Seeing a _slight_ performance hit but nothing too significant. My guess is this is caused by the lock on the `DirectiveIndex`. I'm planning on making the DirectiveIndex read-only  in a future PR (the one that adds docs support) to remove the need for the lock so I'd prefer to land this as-is and make sure we address any performance hit then.

```shell
$ swift run make-test-bundle --output ~/Downloads/test-bundle.docc --sizeFactor 10
$ swift run --package-path bin/benchmark benchmark measure-commits main 473e21a82e14f475f5c20e4d1aa87ae85619b4d8 --repetitions 5 --docc-arguments convert ~/Downloads/test-bundle.docc/TestFramework/Docs.docc
$ swift run --package-path bin/benchmark benchmark diff benchmark-main.json benchmark-473e21a82e14f475f5c20e4d1aa87ae85619b4d8.json


┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Metric                                   │ Change          │ main                 │ 473e21a82e14f475f5c20e4d1aa87ae85619b4d8 │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Duration for 'bundle-registration'       │ no change¹,²    │ 4.056 sec            │ 4.141 sec                                │
│ Duration for 'convert-total-time'        │ +2.321%³,⁴      │ 5.738 sec            │ 5.871 sec                                │
│ Duration for 'documentation-processing'  │ no change⁵      │ 1.516 sec            │ 1.563 sec                                │
│ Duration for 'finalize-navigation-index' │ no change⁶      │ 0.089 sec            │ 0.088 sec                                │
│ Peak memory footprint                    │ no change⁷,⁸    │ 428.9 MB             │ 432.6 MB                                 │
│ Data subdirectory size                   │ no change       │ 111.7 MB             │ 111.7 MB                                 │
│ Index subdirectory size                  │ no change       │ 994 KB               │ 994 KB                                   │
│ Total DocC archive size                  │ no change       │ 117.8 MB             │ 117.8 MB                                 │
│ Topic Anchor Checksum                    │ no change       │ c8f03d4e81da8e3b6d7f │ c8f03d4e81da8e3b6d7fb7f63c771147         │
│ Topic Graph Checksum                     │ no change       │ edce22bce4f8de475a5e │ edce22bce4f8de475a5e58a2e45dcdc0         │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Reply:
Ok, sounds good. Should we consider landing the performance fix as part of this PR though, otherwise we'll incur a regression?